helix-core/src/main/java/org/apache/helix/task/WorkflowContext.java

```diff
@@ -48,22 +48,26 @@
     LAST_PURGE_TIME,
     StartTime, // TODO this should be named JOB_SCHEDULED_START_TIME, it's not the actual start time of the job
     NAME
-    }
+  }
 
   public static final int NOT_STARTED = -1;
   public static final int UNFINISHED = -1;
 
+  // Note: This field needs to be set if any of the workflow context fields have been changed.
+  // Otherwise, the context will not be written to ZK by the controller.
+  private boolean isModified;
+
   public WorkflowContext(ZNRecord record) {
     super(record);
+    isModified = false;
   }
 
   public void setWorkflowState(TaskState s) {
     String workflowState = _record.getSimpleField(WorkflowContextProperties.STATE.name());
-    if (workflowState == null) {
-      _record.setSimpleField(WorkflowContextProperties.STATE.name(), s.name());
-    } else if (!workflowState.equals(TaskState.FAILED.name()) && !workflowState
-        .equals(TaskState.COMPLETED.name())) {
+    if (workflowState == null || (!workflowState.equals(TaskState.FAILED.name())
```

Comment:
This one could also be consolidated the same way, cutting out `workflowState == null`.

Reply:
We can move the definition of terminal states from Workflow/Job Dispatcher to some constant file. Then let's check whether the state whether in the terminal state.

Reply:
Done.